pkg/didcomm/crypto/jwe/authcrypt/encrypt.go

```diff
@@ -0,0 +1,392 @@
+	CipherText string      `json:"ciphertext,omitempty"`
+}
+
+// jweHeaders are the Protected JWE headers in a map format
+type jweHeaders map[string]string
+
+// Recipient is a recipient of an envelope including the shared encryption key
+type Recipient struct {
+	EncryptedKey string           `json:"encrypted_key,omitempty"`
+	Header       RecipientHeaders `json:"header,omitempty"`
+}
+
+// RecipientHeaders are the recipient headers
+type RecipientHeaders struct {
+	APU string `json:"apu,omitempty"`
+	IV  string `json:"iv,omitempty"`
+	Tag string `json:"tag,omitempty"`
+	KID string `json:"kid,omitempty"`
+	OID string `json:"oid,omitempty"`
+}
+
+// New will create an encrypter instance to 'AuthCrypt' payloads for the given sender and recipients arguments
+// and the encryption alg argument. Possible algorithms supported are:
+// C20P (chacha20-poly1035 ietf)
+// XC20P (xchacha20-poly1035 ietf)
+// The returned crypter contains all the information required to encrypt payloads.
+func New(sender keyPair, recipients []*[chacha20poly1305.KeySize]byte, alg ContentEncryption) (*Crypter, error) {
+	var nonceSize int
+	switch alg {
+	case C20P:
+		nonceSize = chacha20poly1305.NonceSize
+	case XC20P:
+		nonceSize = chacha20poly1305.NonceSizeX
+	default:
+		return nil, errors.New(fmt.Sprintf("encryption algorithm '%s' not supported", alg))
+	}
+	if len(recipients) == 0 {
+		return nil, errors.New("empty recipients keys, must have at least one recipient")
+	}
+	var recipientsKey []*[chacha20poly1305.KeySize]byte
+	recipientsKey = append(recipientsKey, recipients...)
+
+	c := &Crypter{
+		sender,
+		recipientsKey,
+		alg,
+		nonceSize,
+	}
+
+	if !isKeyPairValid(sender) {
+		return nil, errors.New(fmt.Sprintf("sender keyPair not supported, it must have %d bytes keys", chacha20poly1305.KeySize))
+	}
+
+	return c, nil
+}
+
+func isKeyPairValid(kp keyPair) bool {
+	if kp.priv == nil || kp.pub == nil {
+		return false
+	}
+
+	return true
+}
+
+// Encrypt will JWE encode the payload argument for the sender and recipients
+// Using (X)Chacha20 encryption algorithm and Poly1035 authenticator
+func (c *Crypter) Encrypt(payload string) ([]byte, error) {
+	headers := jweHeaders{
+		"typ": "prs.hyperledger.aries-auth-message",
+		"alg": "ECDH-SS+" + string(c.alg) + "KW",
+		"enc": string(c.alg),
+	}
+
+	aad := c.buildAAD()
+	aadEncoded := base64.URLEncoding.EncodeToString(aad)
+
+	encHeaders, err := json.Marshal(headers)
+	if err != nil {
+		return nil, err
+	}
+	// build the Payload's AAD string
+	pldAad := base64.URLEncoding.EncodeToString(encHeaders) + "." + aadEncoded
+
+	// generate a new nonce for this encryption
+	nonce := make([]byte, c.nonceSize)
+	_, err = RandReader.Read(nonce)
+	if err != nil {
+		return nil, err
+	}
+	nonceEncoded := base64.URLEncoding.EncodeToString(nonce)
+
+	var cek *[chacha20poly1305.KeySize]byte
+
+	// generate a cek for encryption (it will be treated as a symmetric key)
+	_, cek, err = box.GenerateKey(RandReader)
```

Comment:
This remains fishy to me - you should be able to generate a symmetric key just from reading from secure rand. Using an asymmetric generator is slower and confusing to the reader. Also how do you know the key is the right size?

Reply:
Also suggest using chacha20poly1305.KeySize constant when you generate. This makes it more obvious :).

Reply:
yes.. .refactoring this..

Reply:
done... it's not an asymmetric key.. removed this old logic